src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -88,25 +88,30 @@ pub struct FileType {
 impl FileType {
     /// The filename for this FileType crated by rustc.
     pub fn output_filename(&self, target: &Target, metadata: Option<&str>) -> String {
+        // Check if we have been provided with a seperate filename for the binary in `[[bin]]`
+        // section of Cargo.toml.
+        let filename = match target.get_binary_name() {
+            Some(bin_name) => bin_name,
+            None => target.crate_name(),
+        };
+
         match metadata {
-            Some(metadata) => format!(
-                "{}{}-{}{}",
-                self.prefix,
-                target.crate_name(),
-                metadata,
-                self.suffix
-            ),
-            None => format!("{}{}{}", self.prefix, target.crate_name(), self.suffix),
+            Some(metadata) => format!("{}{}-{}{}", self.prefix, filename, metadata, self.suffix),
+            None => format!("{}{}{}", self.prefix, filename, self.suffix),
         }
     }
 
     /// The filename for this FileType that Cargo should use when "uplifting"
     /// it to the destination directory.
     pub fn uplift_filename(&self, target: &Target) -> String {
+        // for binary crate type, `should_replace_hyphens` will always be false
         let name = if self.should_replace_hyphens {
```

Comment:
Is my assumption that `should_replace_hyphens` will always be `false` for binary crates, correct?

Reply:
I think so, though this seems a little subtle.  Can you test on Windows to see the behavior with PDB files works correctly in the visual studio debugger?  There is a brief explanation [here](https://github.com/rust-lang/cargo/blob/197e657b5bef2964551a127f088da187ccacd30e/src/cargo/core/compiler/build_context/target_info.rs#L392-L405) as to how that works.

Reply:
I would agree that this is a bit subtle and I think it'd be best if this were handled a bit differently where there was an early-return or similar

Reply:
Can you give me a little more background and clues on this? TBH, I'm not a Rust expert (yet). Could you elaborate what you mean by *early return* ?